lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java

```diff
@@ -186,59 +186,26 @@ private static String getBanDurationBreakdown(final Timestamp stamp) {
   private @Nullable String authenticateRegisteredUser(
       final Map<String, String> response, final User user) {
     final String username = user.getUsername();
-    final String errorMessage = ErrorMessages.AUTHENTICATION_FAILED;
     final HashedPassword hashedPassword = database.getUserDao().getPassword(username);
     if (hashedPassword == null) {
-      return errorMessage;
+      return ErrorMessages.AUTHENTICATION_FAILED;
     }
+
     if (RsaAuthenticator.canProcessResponse(response)) {
       return rsaAuthenticator.decryptPasswordForAction(
           response,
           pass -> {
-            final String legacyHashedPassword =
-                response.get(LobbyLoginResponseKeys.HASHED_PASSWORD);
             if (hashedPassword.isBcrypted()) {
-              if (database.getUserDao().login(username, new HashedPassword(pass))) {
-                if (legacyHashedPassword != null
-                    && database.getUserDao().getLegacyPassword(username).value.isEmpty()) {
-
-                  final String email = database.getUserDao().getUserEmailByName(username);
-                  database
-                      .getUserDao()
-                      .updateUser(username, email, new HashedPassword(legacyHashedPassword));
-                  database
-                      .getUserDao()
-                      .updateUser(
-                          username,
-                          email,
-                          new HashedPassword(BCrypt.hashpw(pass, bcryptSaltGenerator.get())));
-                }
-                return null;
-              }
-              return errorMessage;
-            } else if (database
-                .getUserDao()
-                .login(username, new HashedPassword(legacyHashedPassword))) {
-              final String email = database.getUserDao().getUserEmailByName(username);
-              database
-                  .getUserDao()
-                  .updateUser(
-                      username,
-                      email,
-                      new HashedPassword(BCrypt.hashpw(pass, bcryptSaltGenerator.get())));
-              return null;
+              return database.getUserDao().login(username, new HashedPassword(pass))
+                  ? null
+                  : ErrorMessages.AUTHENTICATION_FAILED;
             } else {
-              return errorMessage;
```

Comment:
To clarify:
Are you planning to drop the other password column at some point or keep it for now?

Reply:
Only latest clients will be able to connect to the next lobby. The code removed should then be simply dead code.

With luck we've got it so that any accounts with only md5 encrypted passwords will still be able to login, but looking at the code removed, it doesn't quite look like we'll do the upgrade of password for the case of latest client still using md5.

Push comes to shove we can ask users to use the 'forgot password' feature coming along to reset their password. Bottom line though is the code removed should be dead code.

Reply:
I mean one option would just be to grab all the emails we have of people that don't have an updated password, send them a mail (manually over some mass email service or something) they have to login using the latest stable, otherwise their account will be deleted on October 1st (or other planned upgrade date).
Then we could safely drop the column, rename the current column potentially, and remove all users that don't have an updated password column. They have been warned.

Reply:
> To clarify:
Are you planning to drop the other password column at some point or keep it for now?

There is a select with coelesce on the old password column; it is not clear to me that we can drop it.

Reply:
True, but if we'd just remove all users without a migrated password, the coelesce wouldn't be needed

Reply:
I'm hesitant to remove the column and coelesce. This PR was me going through with a debugger and only removing dead code branches.

For users with an email, they'll be able to use 'forgot password' and we won't need to drop their user row. Otherwise users that have not upgraded will have their game ***freeze on startup***. We've a latent bug in the 3xxx version where if the 'triplea min' version is incremented their game freezes. There is a lot of motivation to get users on 7xxx versions of the game.

Ultimately I suspect users without an email or a bcrypted password would be dropped. For now delaying that decision is good, it's not strictly needed to make that call just yet, but it seems to be coming soon.

Reply:
Ah ok now I get what you are trying to achieve.
I can tell you with high confidence that logging in with a md5-only password won't work. The Md5Crypt challenge code you removed is crucial for this: From what I found out quite some time ago the legacy login system doesn't use Md5Crypt to actually hash passwords before putting them in the database, instead it is used to act as pseudo encryption before actually sending the password over the network.
So in order for the client to authenticate with md5 passwords, the server needs to send the salt of the password stored in the database to the client, which then Md5Crypt-hashes the password using this exact salt and sends the hash over the network which is then compared with the database on the server.

That's why I was so confused initially. For me this PR looks like it removes the critical legacy authentication functions, but doesn't remove everything related to it.

Thoughts?

Reply:
Looking at this code below (LobbyLoginValidator:67), there is no client code that fetches the SALT key from challenge:

```
  private Map<String, String> newMd5CryptAuthenticatorChallenge(final String userName) {
    final Map<String, String> challenge = new HashMap<>();
    final HashedPassword password = database.getUserDao().getLegacyPassword(userName);
    if (password != null && Strings.emptyToNull(password.value) != null) {
      challenge.put(LobbyLoginChallengeKeys.SALT, Md5Crypt.getSalt(password.value));
    }
    return challenge;
  }

```


Looking at this code (LobbyLoginValidator:278), there is no code that 'puts' the HASHED_PASSWORD key into a challenge. All accessors to that property are 'get' and are on the server.

```
   final HashedPassword password =
        new HashedPassword(response.get(LobbyLoginResponseKeys.HASHED_PASSWORD));
    if (RsaAuthenticator.canProcessResponse(response)) {
      return rsaAuthenticator.decryptPasswordForAction(
          response,
          pass -> {
            final HashedPassword newPass =
                new HashedPassword(BCrypt.hashpw(pass, bcryptSaltGenerator.get()));
            if (password.isHashedWithSalt()) {
              database.getUserDao().createUser(username, email, password);
              database.getUserDao().updateUser(username, email, newPass);
            } else {
              database.getUserDao().createUser(username, email, newPass);
            }
            return null;
          });
    }
```

It somewhat seems that we might have a mistake where latest clients cannot do MD5Crypt password authentication to start with. Regardless if that is the case, given all current client, the code removed seem to be impossible code paths when using a current client.


> For me this PR looks like it removes the critical legacy authentication functions, but doesn't remove everything related to it.

In short, it appears all legacy authentication functions are to support 'old' clients. If we assume a latest client with current code, those authentication mechanisms would never have been triggered to start with. Hence, moves to remove those paths.

Reply:
You're right. The current client can't use that code at all, I forgot that we dropped some code on the client side quite some time ago. I even found @ssoloff 's PR https://github.com/triplea-game/triplea/pull/4262

I don't even know which engine version this represents (probably 1.9).

So there's a thing I don't quite understand: If old login handling code is removed which makes it impossible logging in using a legacy client, why keep the HashedPassword wrapper class for example that is only used to handle different types of hashes?
If this is just "step 1" of the removal then I'm ok with it, but I feel like you didn't make clear what comes next exactly in terms of refactoring (the high level explanation with the parallel lobby did help though, explains why the DB structure will be kept).

Reply:
>  If old login handling code is removed which makes it impossible logging in using a legacy client

Assuming I'm correct that these are impossible code paths with a current client, and given we will force a latest client for the 1.10/2.0 release, then it was already impossible (dead code).

>  I'm ok with it, but I feel like you didn't make clear what comes next exactly in terms of refactoring

Great question, TBH I'm not sure. There are some big refactoring changes that could happen, but it's too much for one PR. For now this is taking a slice off of the complexity so the code is a bit easier to work with by removing dead code. With luck password reset will be easier to implement, I spent a lot of time in this code yesterday working on the encryption pieces and eventually concluded the code needed simplification first or otherwise the risk was too high and ultimately wasting time working in an overly complicated area of the code.

It does seem there could be more iterations. With luck those would be more obvious now that this work is done. Getting to that point where such things would be move obvious is perhaps the largest goal here, not necessarily reducing this code to the simplest possible. I am tempted to spend some more time to try and improve things a bit more before picking back up password reset.